### PR TITLE
support numpy fancy indexing for colors and data features

### LINKED ISSUE #101: Implement fancy indexing for GraphicFeatures
We need to implement fancy indexing for indexable graphic features so that these things are possible:

```python
line = plot.add_line(array, colors=(0., 0., 0., 0.))

line.colors[array > 5] = "r"
```

Can probably be implemented in `cleanup_slice` by using the mask and getting the indices, `np.where(array > 5)` etc.

### LINKED ISSUE #102: handling numpy int types
probably need to make numpy int scalers into python native types, this is annoying. Maybe implement in `cleanup_slice()`

```python
for start_ix in ca_trial_range_ixs[:, 0]:
    start_ix = int(start_ix)
    plot_temporal["cue"].colors[start_ix:(start_ix + cue_frame_calcium)] = np.repeat(np.array([Color("w")]), 30, axis=0)
```

ca_trial_range_ixs is an array like this:

```python
array([[   0,  290],
       [ 290,  580],
       [ 580,  870],
       [ 870, 1160],
       [1160, 1450],
       [1450, 1740],
       [1740, 2030],
       [2030, 2320],
       [2320, 2610],
       [2610, 2900],
       [2900, 3190],
       [3190, 3480],
       [3480, 3770],
       [3770, 4060],
       [4060, 4350],
       [4350, 4640],
       [4640, 4930],
       [4930, 5220],
       [5220, 5510],
       [5510, 5800],
       [5800, 6090],
       [6090, 6380],
       [6380, 6670],
       [6670, 6960],
       [6960, 7250],
       [7250, 7540],
       [7540, 7830],
       [7830, 8120],
       [8120, 8410]])
```

### PR DESCRIPTION
**Pull request by kushalkolar**
closes #101, #102 

@clewis7 can you test?

### CODE REVIEW

File: fastplotlib/graphics/features/_base.py [hunk 1]
```diff
@@ -197,13 +203,54 @@ def cleanup_slice(key: Union[int, slice], upper_bound) -> Union[slice, int]:
...
 
     step = key.step
     if step is None:
         step = 1
 
     return slice(start, stop, step)
+    # return slice(int(start), int(stop), int(step))
+
+
+def cleanup_array_slice(key: np.ndarray, upper_bound) -> np.ndarray:
```
Comment (clewis7) [inline fastplotlib/graphics/features/_base.py:216]: LGTM, basically just allows a user to pass in a 1D boolean array or integer array for indices right?

Comment (clewis7) [reply fastplotlib/graphics/features/_base.py:216]: My only thought is what happens when you pass in a boolean array that is not the same length as your data/colors? Does that mean you will only every index from the beginning to how big your boolean array is?

For example:

graphic = line collection with 25 lines
```
bool_array = [True, True, True, False, False]
graphic.data[bool_array]
```

Would that give you only the first three lines in the collection? Or lines 1-3 and 6-25 as well?

Comment (clewis7) [reply fastplotlib/graphics/features/_base.py:216]: Will test out!!

**Comment by clewis7** [comment]
```
from fastplotlib import Plot
import numpy as np

xs = np.linspace(-10, 10, 100)
ys = np.sin(xs)
sine = np.dstack([xs, ys])[0]

ys = np.cos(xs) + 5
cosine = np.dstack([xs, ys])[0]

a = 0.5
ys = np.sinc(xs) * 3 + 8
sinc = np.dstack([xs, ys])[0]

plot = Plot()

coll = plot.add_line_collection(data=[sine, cosine, sinc], colors=["Orange", "Blue", "Gray"])

plot.show()
plot.auto_scale()
```

Screenshot_20230419_091939


```
key = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 67, 19])

coll[0].colors[key] = "Red"

coll[1].data[key] = np.array([[-10, 1, 0]])

key2 = np.array([True, False, True, False, True, True, True, True])

coll[2].data[key2] = np.array([[5, 1, 2]])

coll[2].colors[key2] = "Green"
```

Screenshot_20230419_091914

Seems to work great!!

File: fastplotlib/graphics/features/_base.py (hunk 1 above)
Comment (kushalkolar) [reply fastplotlib/graphics/features/_base.py:216]: yes, so you can do `sine_grahic.colors[ys > 5] = "blue"`

**Comment by kushalkolar** [comment]
@clewis7 can you also test for line stack? And that fancy indexing works on line stack

```python
line_stack[(val > 5) & (val < 10)].colors = "cyan"
```

**Comment by clewis7** [comment]
> @clewis7 can you also test for line stack? And that fancy indexing works on line stack
> 
> ```python
> line_stack[(val > 5) & (val < 10)].colors = "cyan"
> ```

what are you defining as `val` here?

**Comment by clewis7** [comment]
```
from fastplotlib import Plot
import numpy as np

xs = np.linspace(-10, 10, 100)
sys = np.sin(xs)
sine = np.dstack([xs, sys])[0]

cys = np.cos(xs) + 5
cosine = np.dstack([xs, cys])[0]

a = 0.5
snys = np.sinc(xs) * 3 + 8
sinc = np.dstack([xs, snys])[0]

plot = Plot()

stack = plot.add_line_stack(data=[sine, cosine, sinc], colors=["Orange", "Blue", "Gray"])

plot.show()
plot.auto_scale()
```
Screen Shot 2023-04-20 at 10 27 14 AM

```
stack[1].colors[(cys > 5) & (cys < 10)] = "cyan"
stack[0].colors[(xs > 3) & (xs < 10)] = "red"
```
Screen Shot 2023-04-20 at 10 27 37 AM


Works with stack as well!

**Comment by kushalkolar** [comment]
if there are bugs I'll discover them over the coming days of complex plots!

**Pull request merged**